components/calendar/src/chinese_based.rs

```diff
@@ -0,0 +1,331 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and traits for use in the Chinese traditional lunar calendar,
+//! as well as in related and derived calendars such as the Korean and Vietnamese lunar calendars.
+
+use crate::{
+    astronomy::{Astronomical, Location, MEAN_SYNODIC_MONTH, MEAN_TROPICAL_YEAR},
+    calendar_arithmetic::{ArithmeticDate, CalendarArithmetic},
+    helpers::{adjusted_rem_euclid, i64_to_i32, quotient, I32Result},
+    rata_die::RataDie,
+    types::Moment,
+    Calendar, Date,
+};
+
+/// For an example of how to use this trait, see `impl ChineseBased<Chinese>` in [`Chinese`].
+///
+/// The trait ChineseBased is used by Chinese-based calendars to perform computations shared by such calendars.
+/// To do so, calendars should implement `fn location` by providing a location at which observations of the
+/// new moon are recorded (which may change over time), with the most important part being the time zone
+/// offset (longitude, latitude, and elevation are not relevant for these particular calculations);
+/// define `const EPOCH` as a `RataDie` marking the start date of the era of the Calendar for internal use,
+/// which may not accurately reflect how years or eras are marked traditionally or how they will be seen
+/// by end-users; and implement `fn new_chinese_based_date` by taking a year, month, and day, and
+/// returning a Date of the relevant Calendar type.
+pub(crate) trait ChineseBased<C: ChineseBased<C> + CalendarArithmetic> {
```

Comment:
Saying that the trait you're defining `ChineseBased<C>` depends on a `C` such that it implements `ChineseBased<C>` seems recursive & thus redundant. Could you replace that with `C: Calendar + ...` and have it still work?

Reply:
That is circular, I've replaced it with `C: CalendarArithmetic` instead since `CalendarArithmetic` requires `Calendar`.